File: deps/src/knuth-bendix.cpp
//
// Semigroups.jl
// Copyright (C) 2026, James W. Swent
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// CRITICAL: libsemigroups_julia.hpp MUST be included first (fmt consteval fix)
#include "libsemigroups_julia.hpp"

#include <libsemigroups/cong-common-helpers.hpp>
#include <libsemigroups/knuth-bendix-class.hpp>
#include <libsemigroups/knuth-bendix-helpers.hpp>
#include <libsemigroups/presentation.hpp>
#include <libsemigroups/word-graph.hpp>

#include <jlcxx/array.hpp>

#include <chrono>
#include <cstddef>
#include <cstdint>
#include <iterator>
#include <string>
#include <type_traits>
#include <utility>
#include <vector>

namespace jlcxx {
  template <>
  struct IsMirroredType<
      libsemigroups::detail::KnuthBendixImpl<libsemigroups::detail::RewriteTrie,
                                             libsemigroups::ShortLexCompare>>
      : std::false_type {};

  template <>
  struct IsMirroredType<
      libsemigroups::KnuthBendix<libsemigroups::word_type,
                                 libsemigroups::detail::RewriteTrie,
                                 libsemigroups::ShortLexCompare>>
      : std::false_type {};

  template <>
  struct SuperType<
      libsemigroups::detail::KnuthBendixImpl<libsemigroups::detail::RewriteTrie,
                                             libsemigroups::ShortLexCompare>> {
    using type = libsemigroups::detail::CongruenceCommon;
  };

  template <>
  struct SuperType<
      libsemigroups::KnuthBendix<libsemigroups::word_type,
                                 libsemigroups::detail::RewriteTrie,
                                 libsemigroups::ShortLexCompare>> {
    using type = libsemigroups::detail::KnuthBendixImpl<
        libsemigroups::detail::RewriteTrie,
        libsemigroups::ShortLexCompare>;
  };
}  // namespace jlcxx

namespace libsemigroups_julia {

  void define_knuth_bendix(jl::Module& m) {
    using libsemigroups::congruence_kind;
    using libsemigroups::Presentation;
    using libsemigroups::word_type;

    using CongruenceCommon = libsemigroups::detail::CongruenceCommon;
    using KBImpl           = libsemigroups::detail::KnuthBendixImpl<
        libsemigroups::detail::RewriteTrie,
        libsemigroups::ShortLexCompare>;
    using KB = libsemigroups::KnuthBendix<word_type,
                                          libsemigroups::detail::RewriteTrie,
                                          libsemigroups::ShortLexCompare>;

    ////////////////////////////////////////////////////////////////////////
    // overlap enum
    ////////////////////////////////////////////////////////////////////////

    m.add_bits<KB::options::overlap>("overlap", jl::julia_type("CppEnum"));
    m.set_const("overlap_ABC", KB::options::overlap::ABC);
    m.set_const("overlap_AB_BC", KB::options::overlap::AB_BC);
    m.set_const("overlap_MAX_AB_BC", KB::options::overlap::MAX_AB_BC);

    ////////////////////////////////////////////////////////////////////////
    // Type registration
    ////////////////////////////////////////////////////////////////////////

    m.add_type<KBImpl>("KnuthBendixImplRewriteTrie",
                       jlcxx::julia_base_type<CongruenceCommon>());
    auto type = m.add_type<KB>("KnuthBendixRewriteTrie",
                               jlcxx::julia_base_type<KBImpl>());

    ////////////////////////////////////////////////////////////////////////
    // Constructors
    ////////////////////////////////////////////////////////////////////////

    type.constructor<congruence_kind, Presentation<word_type> const&>();
    type.constructor<KB const&>();  // copy ctor
    type.method("init!", [](KB& self) -> KB& { return self.init(); });
    type.method("init!",
                [](KB&                            self,
                   congruence_kind                knd,
                   Presentation<word_type> const& p) -> KB& {
                  return self.init(knd, p);
                });

    ////////////////////////////////////////////////////////////////////////
    // Settings (getter / setter with DISTINCT names)
    ////////////////////////////////////////////////////////////////////////

    // max_pending_rules
    type.method("max_pending_rules", [](KB const& self) -> size_t {
      return self.max_pending_rules();
    });
    type.method("set_max_pending_rules!",
                [](KB& self, size_t val) { self.max_pending_rules(val); });

    // check_confluence_interval
    type.method("check_confluence_interval", [](KB const& self) -> size_t {
      return self.check_confluence_interval();
    });
    type.method("set_check_confluence_interval!", [](KB& self, size_t val) {
      self.check_confluence_interval(val);
    });

    // max_overlap
    type.method("max_overlap",
                [](KB const& self) -> size_t { return self.max_overlap(); });
    type.method("set_max_overlap!",
                [](KB& self, size_t val) { self.max_overlap(val); });

    // max_rules
    type.method("max_rules",
                [](KB const& self) -> size_t { return self.max_rules(); });
    type.method("set_max_rules!",
                [](KB& self, size_t val) { self.max_rules(val); });

    // overlap_policy
    type.method("overlap_policy", [](KB const& self) -> KB::options::overlap {
      return self.overlap_policy();
    });
    type.method("set_overlap_policy!", [](KB& self, KB::options::overlap val) {
      self.overlap_policy(val);
    });

    ////////////////////////////////////////////////////////////////////////
    // Query methods
    ////////////////////////////////////////////////////////////////////////

    type.method("number_of_active_rules", [](KB const& self) -> size_t {
      return self.number_of_active_rules();
    });
    type.method("number_of_inactive_rules", [](KB const& self) -> size_t {
      return self.number_of_inactive_rules();
    });
    type.method("number_of_pending_rules", [](KB const& self) -> size_t {
      return self.number_of_pending_rules();
    });
    type.method("total_rules",
                [](KB const& self) -> size_t { return self.total_rules(); });

    type.method("confluent", [](KB& self) -> bool { return self.confluent(); });
    type.method("confluent_known",
                [](KB const& self) -> bool { return self.confluent_known(); });

    type.method("number_of_classes",
                [](KB& self) -> uint64_t { return self.number_of_classes(); });

    type.method("kind",
                [](KB const& self) -> congruence_kind { return self.kind(); });
    type.method("number_of_generating_pairs", [](KB const& self) -> size_t {
      return self.number_of_generating_pairs();
    });
    type.method("generating_pairs",
                [](KB const& self) -> std::vector<word_type> {
                  auto const& pairs = self.generating_pairs();
                  return std::vector<word_type>(pairs.begin(), pairs.end());
                });

    // presentation — return by copy
    type.method("presentation", [](KB const& self) -> Presentation<word_type> {
      return self.presentation();
    });

    ////////////////////////////////////////////////////////////////////////
    // Word operations (ArrayRef<size_t> for word inputs)
    ////////////////////////////////////////////////////////////////////////

    // reduce (triggers full enumeration)
    m.method("kb_reduce", [](KB& self, jlcxx::ArrayRef<size_t> w) -> word_type {
      word_type input(w.begin(), w.end());
      return libsemigroups::knuth_bendix::reduce(self, input);
    });

    // reduce_no_run (no enumeration)
    m.method("kb_reduce_no_run",
             [](KB const& self, jlcxx::ArrayRef<size_t> w) -> word_type {
               word_type input(w.begin(), w.end());
               return libsemigroups::knuth_bendix::reduce_no_run(self, input);
             });

    // contains (triggers full enumeration)
    m.method("kb_contains",
             [](KB&                     self,
                jlcxx::ArrayRef<size_t> u,
                jlcxx::ArrayRef<size_t> v) -> bool {
               word_type uw(u.begin(), u.end());
               word_type vw(v.begin(), v.end());
               return libsemigroups::knuth_bendix::contains(self, uw, vw);
             });

    // currently_contains (no enumeration, returns tril)
    m.method("kb_currently_contains",
             [](KB const&               self,
                jlcxx::ArrayRef<size_t> u,
                jlcxx::ArrayRef<size_t> v) -> libsemigroups::tril {
               word_type uw(u.begin(), u.end());
               word_type vw(v.begin(), v.end());
               return libsemigroups::knuth_bendix::currently_contains(
                   self, uw, vw);
             });

    // add_generating_pair!
    m.method(
        "kb_add_generating_pair!",
        [](KB& self, jlcxx::ArrayRef<size_t> u, jlcxx::ArrayRef<size_t> v) {
          word_type uw(u.begin(), u.end());
          word_type vw(v.begin(), v.end());
          libsemigroups::knuth_bendix::add_generating_pair(self, uw, vw);
        });

    ////////////////////////////////////////////////////////////////////////
    // Rules access
    ////////////////////////////////////////////////////////////////////////

    // active_rules — collect the range to a flat vector<word_type>
    // (even indices = lhs, odd indices = rhs)
    // active_rules() returns an rx range — use .at_end()/.get()/.next()
    m.method("kb_active_rules", [](KB& self) -> std::vector<word_type> {
      std::vector<word_type> result;
      auto                   range = self.active_rules();
      while (!range.at_end()) {
        auto const& pair = range.get();
        result.push_back(pair.first);
        result.push_back(pair.second);
        range.next();
      }
      return result;
    });

    ////////////////////////////////////////////////////////////////////////
    // Graph access
    ////////////////////////////////////////////////////////////////////////

    // gilman_graph — return by const reference (large stable data)
    type.method("gilman_graph",
                [](KB& self) -> libsemigroups::WordGraph<uint32_t> const& {
                  return self.gilman_graph();
                });

    // gilman_graph_node_labels — collect to vector of vectors by copy
    type.method("gilman_graph_node_labels",
                [](KB& self) -> std::vector<word_type> {
                  return self.gilman_graph_node_labels();
                });

    ////////////////////////////////////////////////////////////////////////
    // Display
    ////////////////////////////////////////////////////////////////////////

    type.method("to_human_readable_repr", [](KB& self) -> std::string {
      return libsemigroups::to_human_readable_repr(self);
    });

    ////////////////////////////////////////////////////////////////////////
    // Free functions (knuth_bendix:: namespace)
    ////////////////////////////////////////////////////////////////////////

    // by_overlap_length! (mutating)
    m.method("kb_by_overlap_length!", [](KB& self) {
      libsemigroups::knuth_bendix::by_overlap_length(self);
    });

    // is_reduced
    m.method("kb_is_reduced", [](KB& self) -> bool {
      return libsemigroups::knuth_bendix::is_reduced(self);
    });

    // redundant_rule — takes Presentation<word_type> and a timeout in
    // nanoseconds (int64_t), returns an index into p.rules (0-based).
    // Returns p.rules.size() if no redundant rule was found.
    m.method("kb_redundant_rule",
             [](Presentation<word_type> const& p, int64_t ns) -> size_t {
               auto it = libsemigroups::knuth_bendix::redundant_rule(
                   p, std::chrono::nanoseconds(ns));
               return static_cast<size_t>(std::distance(p.rules.cbegin(), it));
             });

    // normal_forms — collect to vector<word_type>
    // normal_forms() returns an rx-style range — use .at_end()/.get()/.next()
    m.method("kb_normal_forms", [](KB& self) -> std::vector<word_type> {
      std::vector<word_type> result;
      auto range = libsemigroups::knuth_bendix::normal_forms(self);
      while (!range.at_end()) {
        result.push_back(range.get());
        range.next();
      }
      return result;
    });

    // non_trivial_classes — takes two KB objects
    m.method("kb_non_trivial_classes",
             [](KB& kb1, KB& kb2) -> std::vector<std::vector<word_type>> {
               return libsemigroups::knuth_bendix::non_trivial_classes(kb1,
                                                                       kb2);
             });

    m.method("kb_partition",
             [](KB& self, jlcxx::ArrayRef<jl_value_t*> words)
                 -> std::vector<std::vector<word_type>> {
               std::vector<word_type> input;
               input.reserve(words.size());
               for (jl_value_t* word_value : words) {
                 auto word = jlcxx::ArrayRef<size_t>(
                     reinterpret_cast<jl_array_t*>(word_value));
                 input.emplace_back(word.begin(), word.end());
               }
               return libsemigroups::congruence_common::partition(
                   self, input.begin(), input.end());
             });
  }

}  // namespace libsemigroups_julia

File: src/knuth-bendix.jl
# Copyright (c) 2026, James W. Swent
#
# Distributed under the terms of the GPL license version 3.

"""
knuth-bendix.jl - KnuthBendix wrapper (Layer 2 + 3)
"""

# ============================================================================
# Type alias and constants
# ============================================================================

"""
    KnuthBendix

Type for Knuth-Bendix completion algorithm objects.

A `KnuthBendix` object computes a confluent rewriting system from a
[`Presentation`](@ref Semigroups.Presentation) and a
[`congruence_kind`](@ref Semigroups.congruence_kind). It is a subtype of
[`Runner`](@ref Semigroups.Runner), so all runner methods (`run!`,
`run_for!`, `finished`, `timed_out`, etc.) work on `KnuthBendix` objects.

# Constructors

    KnuthBendix(kind::congruence_kind, p::Presentation) -> KnuthBendix
    KnuthBendix(other::KnuthBendix) -> KnuthBendix

The first form constructs a `KnuthBendix` from a congruence kind and a
presentation. The second form copies an existing `KnuthBendix`.

!!! warning "v1 limitation"
    v1 of Semigroups.jl binds `KnuthBendix<word_type, RewriteTrie>` only.
"""
const KnuthBendix = LibSemigroups.KnuthBendixRewriteTrie

"""
    overlap_ABC

Overlap policy constant: consider overlaps of the form ABC.
"""
const overlap_ABC = LibSemigroups.overlap_ABC

"""
    overlap_AB_BC

Overlap policy constant: consider overlaps of the form AB_BC.
"""
const overlap_AB_BC = LibSemigroups.overlap_AB_BC

"""
    overlap_MAX_AB_BC

Overlap policy constant: consider overlaps of the form MAX_AB_BC.
"""
const overlap_MAX_AB_BC = LibSemigroups.overlap_MAX_AB_BC

# ============================================================================
# Initialization
# ============================================================================

"""
    init!(kb::KnuthBendix) -> KnuthBendix
    init!(kb::KnuthBendix, kind::congruence_kind, p::Presentation) -> KnuthBendix

Re-initialize `kb`.

The one-argument form clears the presentation, generating pairs, settings,
and rewriting data, putting `kb` back into the same state as a newly
default-constructed `KnuthBendix`. The three-argument form reinitializes
`kb` as if it had just been constructed from `kind` and `p`.
"""
function init!(kb::KnuthBendix)
    @wrap_libsemigroups_call LibSemigroups.init!(kb)
    return kb
end

function init!(kb::KnuthBendix, kind::congruence_kind, p::Presentation)
    @wrap_libsemigroups_call LibSemigroups.init!(kb, kind, p)
    return kb
end

# ============================================================================
# Settings — getter / setter pairs
# ============================================================================

"""
    max_pending_rules(kb::KnuthBendix) -> Int

Return the current maximum number of pending rules.
"""
max_pending_rules(kb::KnuthBendix) = Int(LibSemigroups.max_pending_rules(kb))

"""
    max_pending_rules!(kb::KnuthBendix, n::Integer) -> KnuthBendix

Set the maximum number of pending rules. Returns `kb` for chaining.
"""
function max_pending_rules!(kb::KnuthBendix, n::Integer)
    LibSemigroups.set_max_pending_rules!(kb, UInt(n))
    return kb
end

"""
    check_confluence_interval(kb::KnuthBendix) -> Int

Return the current check-confluence interval.
"""
check_confluence_interval(kb::KnuthBendix) =
    Int(LibSemigroups.check_confluence_interval(kb))

"""
    check_confluence_interval!(kb::KnuthBendix, n::Integer) -> KnuthBendix

Set the check-confluence interval. Returns `kb` for chaining.
"""
function check_confluence_interval!(kb::KnuthBendix, n::Integer)
    LibSemigroups.set_check_confluence_interval!(kb, UInt(n))
    return kb
end

"""
    max_overlap(kb::KnuthBendix) -> Int

Return the current maximum overlap size.
"""
max_overlap(kb::KnuthBendix) = Int(LibSemigroups.max_overlap(kb))

"""
    max_overlap!(kb::KnuthBendix, n::Integer) -> KnuthBendix

Set the maximum overlap size. Returns `kb` for chaining.
"""
function max_overlap!(kb::KnuthBendix, n::Integer)
    LibSemigroups.set_max_overlap!(kb, UInt(n))
    return kb
end

"""
    max_rules(kb::KnuthBendix) -> Int

Return the current maximum number of rules.
"""
max_rules(kb::KnuthBendix) = Int(LibSemigroups.max_rules(kb))

"""
    max_rules!(kb::KnuthBendix, n::Integer) -> KnuthBendix

Set the maximum number of rules. Returns `kb` for chaining.
"""
function max_rules!(kb::KnuthBendix, n::Integer)
    LibSemigroups.set_max_rules!(kb, UInt(n))
    return kb
end

"""
    overlap_policy(kb::KnuthBendix)

Return the current overlap policy.
"""
overlap_policy(kb::KnuthBendix) = LibSemigroups.overlap_policy(kb)

"""
    overlap_policy!(kb::KnuthBendix, val) -> KnuthBendix

Set the overlap policy. Returns `kb` for chaining.
"""
function overlap_policy!(kb::KnuthBendix, val)
    LibSemigroups.set_overlap_policy!(kb, val)
    return kb
end

# ============================================================================
# Query methods
# ============================================================================

"""
    number_of_active_rules(kb::KnuthBendix) -> Int

Return the number of active rules in the rewriting system.
"""
number_of_active_rules(kb::KnuthBendix) = Int(LibSemigroups.number_of_active_rules(kb))

"""
    number_of_inactive_rules(kb::KnuthBendix) -> Int

Return the number of inactive rules in the rewriting system.
"""
number_of_inactive_rules(kb::KnuthBendix) = Int(LibSemigroups.number_of_inactive_rules(kb))

"""
    number_of_pending_rules(kb::KnuthBendix) -> Int

Return the number of pending rules in the rewriting system.
"""
number_of_pending_rules(kb::KnuthBendix) = Int(LibSemigroups.number_of_pending_rules(kb))

"""
    total_rules(kb::KnuthBendix) -> Int

Return the total number of rules (active + inactive + pending).
"""
total_rules(kb::KnuthBendix) = Int(LibSemigroups.total_rules(kb))

"""
    confluent(kb::KnuthBendix) -> Bool

Check if the rewriting system is confluent. May trigger a run.
"""
confluent(kb::KnuthBendix) = LibSemigroups.confluent(kb)

"""
    confluent_known(kb::KnuthBendix) -> Bool

Check if confluence status is known without triggering a run.
"""
confluent_known(kb::KnuthBendix) = LibSemigroups.confluent_known(kb)

"""
    number_of_classes(kb::KnuthBendix) -> UInt64

Return the number of congruence classes. May return `POSITIVE_INFINITY`.
"""
number_of_classes(kb::KnuthBendix) = LibSemigroups.number_of_classes(kb)

"""
    kind(kb::KnuthBendix) -> congruence_kind

Return the congruence kind (twosided or onesided).
"""
kind(kb::KnuthBendix) = LibSemigroups.kind(kb)

"""
    number_of_generating_pairs(kb::KnuthBendix) -> Int

Return the number of generating pairs that have been added.
"""
number_of_generating_pairs(kb::KnuthBendix) =
    Int(LibSemigroups.number_of_generating_pairs(kb))

"""
    generating_pairs(kb::KnuthBendix) -> Vector{Tuple{Vector{Int}, Vector{Int}}}

Return the generating pairs of `kb`.

Words are returned as 1-based `Vector{Int}` letter indices.
"""
function generating_pairs(kb::KnuthBendix)
    flat = LibSemigroups.generating_pairs(kb)
    result = Tuple{Vector{Int},Vector{Int}}[]
    for i = 1:2:length(flat)
        push!(result, (_word_from_cpp(flat[i]), _word_from_cpp(flat[i+1])))
    end
    return result
end

"""
    presentation(kb::KnuthBendix) -> Presentation

Return a copy of the presentation used by `kb`.
"""
presentation(kb::KnuthBendix) = LibSemigroups.presentation(kb)

# ============================================================================
# Rules access
# ============================================================================

"""
    active_rules(kb::KnuthBendix) -> Vector{Tuple{Vector{Int}, Vector{Int}}}

Return all active rules as a vector of `(lhs, rhs)` tuples.

Words are returned as 1-based `Vector{Int}` letter indices.
"""
function active_rules(kb::KnuthBendix)
    flat = @wrap_libsemigroups_call LibSemigroups.kb_active_rules(kb)
    result = Tuple{Vector{Int},Vector{Int}}[]
    for i = 1:2:length(flat)
        push!(result, (_word_from_cpp(flat[i]), _word_from_cpp(flat[i+1])))
    end
    return result
end

# ============================================================================
# Graph access
# ============================================================================

"""
    gilman_graph(kb::KnuthBendix) -> WordGraph

Return the Gilman graph of the confluent rewriting system.
"""
@cxxdereference gilman_graph(kb::KnuthBendix) = LibSemigroups.gilman_graph(kb)

"""
    gilman_graph_node_labels(kb::KnuthBendix) -> Vector{Vector{Int}}

Return the node labels of the Gilman graph as 1-based words.
"""
@cxxdereference function gilman_graph_node_labels(kb::KnuthBendix)
    labels = LibSemigroups.gilman_graph_node_labels(kb)
    return [_word_from_cpp(w) for w in labels]
end

# ============================================================================
# Base.* overloads
# ============================================================================

"""
    Base.length(kb::KnuthBendix) -> UInt64

Return the number of congruence classes. Equivalent to
[`number_of_classes`](@ref).
"""
Base.length(kb::KnuthBendix) = number_of_classes(kb)

"""
    Base.show(io::IO, kb::KnuthBendix)

Print a human-readable representation of `kb`.
"""
function Base.show(io::IO, kb::KnuthBendix)
    print(io, LibSemigroups.to_human_readable_repr(kb))
end

"""
    Base.copy(kb::KnuthBendix) -> KnuthBendix

Create an independent copy of `kb`.
"""
Base.copy(kb::KnuthBendix) = LibSemigroups.KnuthBendixRewriteTrie(kb)

Base.deepcopy_internal(kb::KnuthBendix, ::IdDict) = LibSemigroups.KnuthBendixRewriteTrie(kb)

# ============================================================================
# Free functions
# ============================================================================

"""
    by_overlap_length!(kb::KnuthBendix) -> KnuthBendix

Run Knuth-Bendix by overlap length. Mutating.
"""
function by_overlap_length!(kb::KnuthBendix)
    @wrap_libsemigroups_call LibSemigroups.kb_by_overlap_length!(kb)
    return kb
end

"""
    is_reduced(kb::KnuthBendix) -> Bool

Check if the rewriting system is reduced.
"""
is_reduced(kb::KnuthBendix) = @wrap_libsemigroups_call LibSemigroups.kb_is_reduced(kb)

"""
    redundant_rule(p::Presentation, timeout::TimePeriod) -> Union{Int, Nothing}

Find a redundant rule in `p` using Knuth-Bendix, with the given timeout.

Returns the 1-based rule-pair index of a redundant rule, or `nothing` if
no redundant rule is found within the timeout.
"""
function redundant_rule(p::Presentation, timeout::TimePeriod)
    ns = convert(Nanosecond, timeout)
    idx =
        @wrap_libsemigroups_call LibSemigroups.kb_redundant_rule(p, Int64(Dates.value(ns)))
    n_flat = 2 * number_of_rules(p)
    # If idx == n_flat, no redundant rule was found
    if idx >= n_flat
        return nothing
    end
    # Convert from 0-based flat index to 1-based rule-pair index
    return div(Int(idx), 2) + 1
end
